### ISSUE TITLE
Unnecessary copy being performed when calling flat / moddims

### ISSUE THREAD

**Issue by pavanky** [arrayfire/arrayfire#1101]
Here is some code to reproduce the problem:
``` cpp
    array a = randu(5,5);
    array b = a(span, span);
    array c = flat(b);
    printf("%p, %p, %p\n", a.device<float>(), b.device<float>(), c.device<float>());
```

The issue occurs because internally `b` is not treated as an owner and any modifications of b force a copy.

The way to fix this is to check when the indices refer to the entire array and call `af_retain_array` instead calling `createSubArray`.

**Comment by pavanky** [comment]
@FilipeMaia Because of current limitations, I am reverting the behavior of `device` (not `flat`) to perform an additional memory copy. We'll be adding a new function called `rawPtr` or something similar for the next release that will behave as you expected.

This is related to the issues mentioned here: https://github.com/arrayfire/arrayfire/issues/1135

**Comment by pavanky** [comment]
That said, I think we can verify `flat` is not doing a memory copy by checking `deviceMemInfo`.

**Comment by pavanky** [comment]
Fixed in hotfixes-3.2.1
